designer/server/src/main/scala/pl/touk/nussknacker/ui/process/processingtypedata/ProcessingTypeDataProvider.scala

```diff
@@ -34,75 +32,126 @@ trait ProcessingTypeDataProvider[+T, +C] {
       )
     )
 
-  def all(implicit user: LoggedUser): Map[ProcessingType, T]
+  final def forType(processingType: ProcessingType)(implicit user: LoggedUser): Option[T] = allAuthorized
+    .get(processingType)
+    .map(_.getOrElse(throw new UnauthorizedError()))
+
+  def all(implicit user: LoggedUser): Map[ProcessingType, T] = allAuthorized.collect { case (k, Some(v)) =>
+    (k, v)
+  }
+
+  private def allAuthorized(implicit user: LoggedUser): Map[ProcessingType, Option[T]] = state.all.map {
+    case (k, ValueWithPermission(v, AnyUser)) => (k, Some(v))
+    case (k, ValueWithPermission(v, UserWithAccessRightsToCategory(category))) if user.can(category, Permission.Read) =>
+      (k, Some(v))
+    case (k, _) => (k, None)
+  }
 
   // TODO: We should return a generic type that can produce views for users with access rights to certain categories only.
   //       Thanks to that we will be sure that no sensitive data leak
-  def combined: C
+  def combined: C = state.getCombined()
```

Comment:
it's hard to figure out what the "combined: C" is.

Reply:
What exactly do you propose? Should I describe why we combine something in a comment?

Reply:
I think I didn't understand what is `C` before. WDYT about explaining in the `ProcessingTypeDataProvider` trait comment relations between `ProcessingType`, `LoggedUser`, `Data` and `CombinedData`?

Reply:
Great comment. It is much easier to grasp the sense of the trait now.